### PR TITLE
Remove tp config

### PR DESCRIPTION
**Pull request by nilsteampassnet**
The aim is:

- removal of tp.config.php file to rely only on settings stored in DB. This permits to get rid of any potential vulnerability related to the usage of this file in a malicious way
- improving function getCurrentAccessRights() rated class F
- removing some dead (old) functions
- potential fix for #4317
- initiated a refactoring of class Folder
- added new option in Options page for encryptClientServer

### CODE REVIEW

File: install/upgrade_ajax.php [hunk 1]
```diff
@@ -212,6 +213,10 @@ function settingsConsistencyCheck(): array
 define('MIN_MYSQL_VERSION', '8.0.13');
 define('MIN_MARIADB_VERSION', '10.2.1');
 
+// Load config if $SETTINGS not defined
+$configManager = new ConfigManager();
+$SETTINGS = $configManager->getAllSettings();
```
Comment (corentin-soriano) [inline install/upgrade_ajax.php:218]: This comment seems inconsistent because there are no condition.

File: install/upgrade_operations.php [hunk 2]
```diff
@@ -41,13 +42,16 @@
 set_time_limit(600);
 $_SESSION['CPM'] = 1;
 
+// Load config if $SETTINGS not defined
+$configManager = new ConfigManager();
+$SETTINGS = $configManager->getAllSettings();
```
Comment (corentin-soriano) [inline install/upgrade_operations.php:47]: idem

File: install/upgrade_run_3.0.0.php [hunk 3]
```diff
@@ -43,13 +44,16 @@
 set_time_limit(600);
 $_SESSION['CPM'] = 1;
 
+// Load config if $SETTINGS not defined
+$configManager = new ConfigManager();
+$SETTINGS = $configManager->getAllSettings();
```
Comment (corentin-soriano) [inline install/upgrade_run_3.0.0.php:49]: idem

File: install/upgrade_run_3.0.0_fields.php [hunk 4]
```diff
@@ -41,10 +42,13 @@
 set_time_limit(600);
 $_SESSION['CPM'] = 1;
 
+// Load config if $SETTINGS not defined
+$configManager = new ConfigManager();
+$SETTINGS = $configManager->getAllSettings();
```
Comment (corentin-soriano) [inline install/upgrade_run_3.0.0_fields.php:47]: idem

File: install/upgrade_run_3.0.0_files.php [hunk 5]
```diff
@@ -44,14 +45,16 @@
 set_time_limit(600);
 $_SESSION['CPM'] = 1;
 
+// Load config if $SETTINGS not defined
+$configManager = new ConfigManager();
+$SETTINGS = $configManager->getAllSettings();
```
Comment (corentin-soriano) [inline install/upgrade_run_3.0.0_files.php:50]: idem

File: install/upgrade_run_3.0.0_logs.php [hunk 6]
```diff
@@ -41,10 +42,13 @@
 set_time_limit(600);
 $_SESSION['CPM'] = 1;
 
+// Load config if $SETTINGS not defined
+$configManager = new ConfigManager();
+$SETTINGS = $configManager->getAllSettings();
```
Comment (corentin-soriano) [inline install/upgrade_run_3.0.0_logs.php:47]: idem

File: install/upgrade_run_3.0.0_passwords.php [hunk 7]
```diff
@@ -41,11 +42,14 @@
 set_time_limit(600);
 $_SESSION['CPM'] = 1;
 
+// Load config if $SETTINGS not defined
+$configManager = new ConfigManager();
+$SETTINGS = $configManager->getAllSettings();
```
Comment (corentin-soriano) [inline install/upgrade_run_3.0.0_passwords.php:47]: idem

File: install/upgrade_run_3.0.0_suggestions.php [hunk 8]
```diff
@@ -41,11 +42,14 @@
 set_time_limit(600);
 $_SESSION['CPM'] = 1;
 
+// Load config if $SETTINGS not defined
+$configManager = new ConfigManager();
+$SETTINGS = $configManager->getAllSettings();
```
Comment (corentin-soriano) [inline install/upgrade_run_3.0.0_suggestions.php:47]: idem

File: install/upgrade_run_3.0.0_users.php [hunk 9]
```diff
@@ -42,11 +43,14 @@
 set_time_limit(600);
 $_SESSION['CPM'] = 1;
 
+// Load config if $SETTINGS not defined
+$configManager = new ConfigManager();
+$SETTINGS = $configManager->getAllSettings();
```
Comment (corentin-soriano) [inline install/upgrade_run_3.0.0_users.php:48]: idem

File: install/upgrade_run_3.0.php [hunk 10]
```diff
@@ -42,13 +43,16 @@
 set_time_limit(600);
 $_SESSION['CPM'] = 1;
 
+// Load config if $SETTINGS not defined
+$configManager = new ConfigManager();
+$SETTINGS = $configManager->getAllSettings();
```
Comment (corentin-soriano) [inline install/upgrade_run_3.0.php:48]: idem

File: install/upgrade_run_3.1.php [hunk 11]
```diff
@@ -40,13 +41,16 @@
 set_time_limit(600);
 $_SESSION['CPM'] = 1;
 
+// Load config if $SETTINGS not defined
+$configManager = new ConfigManager();
+$SETTINGS = $configManager->getAllSettings();
```
Comment (corentin-soriano) [inline install/upgrade_run_3.1.php:46]: idem

File: pages/tasks.php [hunk 12]
```diff
@@ -304,6 +304,7 @@
                                             </button>
                                         </div>
                                     </div>
+-->
```
Comment (corentin-soriano) [inline pages/tasks.php:307]: This block should be removed if it has become obsolete.

File: sources/admin.queries.php [hunk 13]
```diff
@@ -1988,7 +1926,7 @@
             $SETTINGS['ga_website_name'] = htmlspecialchars_decode($dataReceived['ga_website_name']);
 
             // save change in config file
-            handleConfigFile('update', $SETTINGS, 'ga_website_name', $SETTINGS['ga_website_name']);
+            //handleConfigFile('update', $SETTINGS, 'ga_website_name', $SETTINGS['ga_website_name']);
```
Comment (corentin-soriano) [inline sources/admin.queries.php:1929]: This block should be removed if it has become obsolete.

File: sources/admin.queries.php [hunk 14]
```diff
@@ -2253,13 +2191,15 @@
...
 
+        /*
         // save change in config file
         handleConfigFile(
             'update',
             $SETTINGS,
             $post_field,
             $dataReceived['value']
         );
+        */
```
Comment (corentin-soriano) [inline sources/admin.queries.php:2202]: This block should be removed if it has become obsolete.

File: sources/admin.queries.php [hunk 15]
```diff
@@ -2338,7 +2278,7 @@
         }
 
         // save change in config file
-        handleConfigFile('update', $SETTINGS, 'send_stats', $SETTINGS['send_stats']);
+        //handleConfigFile('update', $SETTINGS, 'send_stats', $SETTINGS['send_stats']);
```
Comment (corentin-soriano) [inline sources/admin.queries.php:2281]: This block should be removed if it has become obsolete.

File: sources/admin.queries.php [hunk 16]
```diff
@@ -2372,7 +2312,7 @@
         }
 
         // save change in config file
-        handleConfigFile('update', $SETTINGS, 'send_statistics_items', $SETTINGS['send_statistics_items']);
+        //handleConfigFile('update', $SETTINGS, 'send_statistics_items', $SETTINGS['send_statistics_items']);
```
Comment (corentin-soriano) [inline sources/admin.queries.php:2315]: This block should be removed if it has become obsolete.

File: sources/identify.php [hunk 17]
```diff
@@ -2355,7 +2355,7 @@ function shouldUserAuthWithOauth2(
     // Security issue without this return if an user auth_type == oauth2 and
     // oauth2 disabled : we can login as a valid user by using hashUserId(username)
     // as password in the login the form.
-    if ((int) $SETTINGS['oauth2_enabled'] !== 1) {
+    if ((int) $SETTINGS['oauth2_enabled'] !== 1 && (string) $userInfo['auth_type'] === 'oauth2') {
```
Comment (corentin-soriano) [inline sources/identify.php:2358]: Could a local user authenticate via OAuth2 for the first time and "break" their account?

File: sources/main.queries.php [hunk 18]
```diff
@@ -1378,7 +1378,7 @@ function sendingStatistics(
         $SETTINGS['send_stats_time'] = time();
 
         // save change in config file
-        handleConfigFile('update', $SETTINGS, 'send_stats_time', (string) $SETTINGS['send_stats_time']);
+        //handleConfigFile('update', $SETTINGS, 'send_stats_time', (string) $SETTINGS['send_stats_time']);
```
Comment (corentin-soriano) [inline sources/main.queries.php:1381]: This block should be removed if it has become obsolete.

File: sources/scheduler.php [hunk 19]
```diff
@@ -83,13 +86,15 @@
...
 
+/*
 if (isset($SETTINGS['rebuild_config_file_task']) === true && empty($SETTINGS['rebuild_config_file_task']) === false) {
     runTask(
         explode(';', $SETTINGS['rebuild_config_file_task']),
         __DIR__.'/../scripts/task_maintenance_rebuild_config_file.php',
         $scheduler
     );
 }
+*/
```
Comment (corentin-soriano) [inline sources/scheduler.php:97]: This block should be removed if it has become obsolete.

File: sources/tasks.queries.php [hunk 20]
```diff
@@ -224,10 +224,12 @@ function loadLastTasksExec_getBadge(string $processLabel): string
             'db' => 'do_maintenance - purge-old-files',
             'task' => 'purge_temporary_files_task',
         ],
+        /*
         'do_maintenance - rebuild-config-file' => [
             'db' => 'do_maintenance - rebuild-config-file',
             'task' => 'rebuild_config_file_task',
         ],
+        */
```
Comment (corentin-soriano) [inline sources/tasks.queries.php:232]: This block should be removed if it has become obsolete.

File: sources/tasks.queries.php [hunk 21]
```diff
@@ -307,6 +309,7 @@ function performTask(string $task, string $phpBinaryPath, string $datetimeFormat
             ]);
 
             break;
+            */
```
Comment (corentin-soriano) [inline sources/tasks.queries.php:312]: This block should be removed if it has become obsolete.

File: vendor/teampassclasses/configmanager/src/ConfigManager.php [hunk 22]
```diff
@@ -28,8 +28,8 @@
  * @see       https://www.teampass.net
  */
 
-use Symfony\Component\HttpFoundation\Session\Session;
 use TeampassClasses\SessionManager\SessionManager;
+use DB;
 class ConfigManager
```
Comment (corentin-soriano) [inline vendor/teampassclasses/configmanager/src/ConfigManager.php:33]: A line break would improve readability.

File: vendor/teampassclasses/folderservices/src/FolderComplexityService.php [hunk 23]
```diff
@@ -0,0 +1,23 @@
...
+
+    public function addComplexity(int $folderId, int $complexity)
+    {
+        DB::insert(prefixTable('misc'), [
+            'type' => 'complex',
+            'intitule' => $folderId,
+            'valeur' => $complexity,
+            'created_at' => time(),
+        ]);
+    }
```
Comment (corentin-soriano) [inline vendor/teampassclasses/folderservices/src/FolderComplexityService.php:22]: Missing documentations and licence here.

File: vendor/teampassclasses/folderservices/src/FolderCreationService.php [hunk 24]
```diff
@@ -0,0 +1,39 @@
...
+
+    public function createFolder(array $params): int
+    {
+        DB::insert(prefixTable('nested_tree'), [
+            'parent_id' => $params['parent_id'],
+            'title' => $params['title'],
+            'personal_folder' => $params['personal_folder'] ?? 0,
+        ]);
+        return DB::insertId();
+    }
```
Comment (corentin-soriano) [inline vendor/teampassclasses/folderservices/src/FolderCreationService.php:33]: Missing documentations and licence here.

File: vendor/teampassclasses/folderservices/src/PermissionChecker.php [hunk 25]
```diff
@@ -0,0 +1,22 @@
...
+        if ($isAdmin) {
+            return true;
+        }
+        $accessibleFolders = $this->getUserAccessibleFolders($userId);
+        return in_array($parentId, $accessibleFolders);
+    }
+
+    private function getUserAccessibleFolders(int $userId): array
+    {
+        return DB::queryFirstColumn('SELECT folder_id FROM accessible_folders WHERE user_id = %i', $userId);
```
Comment (corentin-soriano) [inline vendor/teampassclasses/folderservices/src/PermissionChecker.php:20]: Missing documentations and licence here.

**Review by corentin-soriano — CHANGES_REQUESTED** [review]
Big changes :)
I’ll take some time tomorrow to test and finish the review!

File: vendor/teampassclasses/folderservices/src/FolderComplexityService.php (hunk 23 above)
Comment (nilsteampassnet) [reply vendor/teampassclasses/folderservices/src/FolderComplexityService.php:22]: Under construction but correct

**Review by corentin-soriano — CHANGES_REQUESTED** [review]
It would be a good idea to take the data from the existing `tp.config.php` and overwrite the content in the database for the first time after update (we can auto remove the file after).
During my tests, some settings were retained but others were lost.

I think you're not finished yet, there are still `include_once` of `tp.config.php` config file in some pages.

File: install/install.queries.php [hunk 26]
```diff
@@ -704,14 +687,14 @@ function encryptFollowingDefuse($message, $ascii_key)
                             array('admin', 'pwd_default_length', '14'),
                             array('admin', 'tasks_log_retention_delay', '30'),
                             array('admin', 'oauth2_enabled', '0'),
-                            array('admin', 'oauth2_client_id', ''),
-                            array('admin', 'oauth2_client_secret', ''),
+                            array('admin', 'oauth2_client_id', '', '1'),
+                            array('admin', 'oauth2_client_secret', '', '1'),
                             array('admin', 'oauth2_client_endpoint', ''),
-                            array('admin', 'oauth2_client_token', ''),
+                            array('admin', 'oauth2_client_token', '', '1'),
```
Comment (corentin-soriano) [inline install/install.queries.php:693]: Missing `oauth2_client_urlResourceOwnerDetails` parameter (seems to have been removed also in the admin settings but it is a used setting).

Comment (corentin-soriano) [inline install/install.queries.php:693]: The last `'1'` is for encryption?
`oauth2_client_token` contains the URL of tokeninfo endpoint and not a confidential token.

File: includes/libraries/teampassclasses/configmanager/src/ConfigManager.php [hunk 27]
```diff
@@ -28,34 +28,37 @@
  * @see       https://www.teampass.net
  */
 
- class ConfigManager
- {
-     private $settings;
+use TeampassClasses\SessionManager\SessionManager;
+use DB;
+class ConfigManager
```
[the remark this answers is not in the document]
Comment (nilsteampassnet) [reply includes/libraries/teampassclasses/configmanager/src/ConfigManager.php:33]: Thank you @corentin-soriano for the idea. I didn't face this during my own tests. 
I will work on it and provide a new commit.

File: install/install.queries.php (hunk 26 above)
Comment (nilsteampassnet) [reply install/install.queries.php:693]: No it is a setting that has been added while provinding generic oauth2 other than EntraID.
So there has never been this setting in admin page.
I will add it

File: install/upgrade_run_3.1.php [hunk 28]
```diff
@@ -541,6 +541,45 @@
...
+
+    // Rename the file
+    $newConfigFilePath = $configFilePath . '.bak';
+    if (!rename($configFilePath, $newConfigFilePath)) {
+        // Remove the file
+        unlink($configFilePath);
+    } else {
+        // Remove the file
+        unlink($configFilePath);
+    }
```
Comment (corentin-soriano) [inline install/upgrade_run_3.1.php:579]: Same actions in if/else?

**Review by corentin-soriano — CHANGES_REQUESTED** [review]
This time my tp.config is copied into the database and the file is deleted.
I can see the alert message if the file is still present.

File: install/install.queries.php (hunk 26 above)
Comment (corentin-soriano) [reply install/install.queries.php]: Nice, thank you!

Comment (nilsteampassnet) [reply install/install.queries.php:694]: Yes but as tpconfig file has been removed, it is not used anymore.
I will merge the PR and remove this after.
OK for you?

**Pull request merged**